src/logcounter.cpp

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (c) 2021 Staz Modrzynski
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include "logcounter.hpp"
+
+rtc::LogCounter::LogCounter(plog::Severity severity, const std::string &text, std::chrono::seconds duration) :
+        severity(severity), text(text), duration(duration) {}
+
+rtc::LogCounter& rtc::LogCounter::operator++(int) {
+    std::lock_guard lock(mutex);
+    count++;
+    if (!future) {
+        future = ThreadPool::Instance().schedule(duration, [this]() {
+            int countCopy;
+            {
+                std::lock_guard lock(mutex);
+                countCopy = count;
+                count = 0;
+                future = std::nullopt;
+            }
+            PLOG(severity) << text << ": " << countCopy << " (over " << std::chrono::duration_cast<std::chrono::seconds>(duration).count() << " seconds)";
+        });
+    }
+    return *this;
+}
+
+rtc::LogCounter::~LogCounter() {
+    if (future) {
+        future->wait();
```

Comment:
I know this isn't correct but I don't know how this should be done.

Reply:
You should not wait on the future and actually do nothing, since you don't know whether the thread pool will have run the task. If it has not this will deadlock.

Reply:
Is there a possibility for a race condition here where the counter would be destroyed before the thread pool finishes doing it's thing?

Reply:
Definitely, because both are global objects, so you have no control over the destruction order.

Reply:
So what exactly should be happening here? Isn't there the potential for a segfault if the ThreadPool triggers the log counter after the log counter was destroyed?

Reply:
I think the easiest way to keep the `LogCounter` instance globals is to store the state of `LogCounter` in a structure held by a `shared_ptr`, and pass a `weak_ptr` to the task.

Reply:
What about using `std::enable_shared_from_this` like I just did?

Reply:
You could do that, however `enable_shared_from_this` would require to create every instance of `LogCounter` with a `shared_ptr`, which is a bit cumbersome.

Reply:
So I did it, but I feel like it might be hacky.